secretstores/json/jsonsecretstore.go

```diff
@@ -0,0 +1,202 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package jsonsecretstore
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/dapr/components-contrib/secretstores"
+	"github.com/dapr/dapr/pkg/logger"
+)
+
+const (
+	/* #nosec */
+	enableSecretStoreVariable = "DAPR_ENABLE_JSON_SECRET_STORE"
+)
+
+type jsonSecretStoreMetaData struct {
+	SecretsFile     string `json:"secretsFile"`
+	NestedSeparator string `json:"nestedSeparator"`
+}
+
+type jsonSecretStore struct {
+	secretsFile     string
+	nestedSeparator string
+	currenContext   []string
+	currentPath     string
+	secrets         map[string]string
+	readJSONFileFn  func(secretsFile string) (map[string]interface{}, error)
+	logger          logger.Logger
+}
+
+// NewJSONSecretStore returns a new JSON secret store
+func NewJSONSecretStore(logger logger.Logger) secretstores.SecretStore {
+	return &jsonSecretStore{
+		logger: logger,
+	}
+}
+
+// Init creates a JSON secret store
+func (j *jsonSecretStore) Init(metadata secretstores.Metadata) error {
+	enable := os.Getenv(enableSecretStoreVariable)
+	if enable != "1" {
+		return fmt.Errorf("jsonsecretstore must be explicitly enabled setting %s environment variable value to 1", enableSecretStoreVariable)
+	}
+
+	meta, err := j.getJSONSecretStoreMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	if len(meta.NestedSeparator) == 0 {
+		j.nestedSeparator = ":"
+	}
+
+	if j.readJSONFileFn == nil {
+		j.readJSONFileFn = j.readJSONFile
+	}
+
+	j.secrets = map[string]string{}
+
+	jsonConfig, err := j.readJSONFileFn(meta.SecretsFile)
+	if err != nil {
+		return err
+	}
+
+	j.visitJSONObject(jsonConfig)
+
+	return nil
+}
+
+// GetSecret retrieves a secret using a key and returns a map of decrypted string/string values
+func (j *jsonSecretStore) GetSecret(req secretstores.GetSecretRequest) (secretstores.GetSecretResponse, error) {
+	secretValue, exists := j.secrets[req.Name]
+	if !exists {
+		return secretstores.GetSecretResponse{}, fmt.Errorf("secret %s not found", req.Name)
+	}
+
+	return secretstores.GetSecretResponse{
+		Data: map[string]string{
+			req.Name: secretValue,
```

Comment:
If we take the example in the issue https://github.com/dapr/components-contrib/issues/327
The req.Name will be ConnectionStrings.
and then it should return a map of string, string:
"MySQL": "super secret connection string"

So shouldn't j.Secrets be a nested map
`map[string]map[string]string`
Then when the request we just lookup req.Name from this map and return the map for it.

Reply:
The way it works is that  when it finds many levels it flattens the hierarchy, so the key is composed. In the future, this will help providers such a as: [Dapr.Extensions.Configuration](https://github.com/dapr/dotnet-sdk/tree/master/src/Dapr.Extensions.Configuration)  to map the secrets to an object.

If you just want the simple key / value match it's as a simple as using a json like the following:

```json
{
  "Secret1": "super secret 1",
  "Secret2": "super secret 2"
}
```

Reply:
@cmendible I was referring to something like folloing:
```
{
        "ConnectionStrings": {
            "MySQL": "super secret connection string",
            "MySecret2": "SecretValue"
        }
   }
```


So shouldn't j.Secrets be a nested map
map[string]map[string]string
Then when the request we just lookup req.Name from this map and return the map for it.

Reply:
I see in that case by default you will access each property by key:

```
ConnectionStrings:MySQL
ConnectionStrings:MySQL:MySecret2
```
the model accepts any numbers of levels of nesting, so you can add for instance a 3rd:

```
ConnectionStrings:SQL:SQL
ConnectionStrings:NoSQL:CosmosDb
```

The implementation is a port of the .NET Configuration JSON providers: https://github.com/aspnet/Configuration/blob/master/src/Config.Json/JsonConfigurationFileParser.cs

Reply:
@cmendible Sounds good, that allows more nested usage. 

> > Thanks @cmendible for the contribution, this will be very helpful for demos, samples and dev environments. Could you please add a doc how to PR as well?
> 
> @amanbha Should I open this PR in the [docs](https://github.com/dapr/docs) repo?

Yes, please open a PR in doc repo.

Reply:
@cmendible sounds good, it allows for more nested usage.